src/main/java/net/earthcomputer/clientcommands/command/ItemGroupCommand.java

```diff
@@ -0,0 +1,361 @@
+                        .then(argument("index", integer(0))
+                            .executes(ctx -> removeStack(ctx.getSource(), getString(ctx, "group"), getInteger(ctx, "index")))))
+                    .then(literal("set")
+                        .then(argument("index", integer(0))
+                            .then(argument("itemstack", itemStack())
+                                .then(argument("count", integer(1))
+                                    .executes(ctx -> setStack(ctx.getSource(), getString(ctx, "group"), getInteger(ctx, "index"), getItemStackArgument(ctx, "itemstack").createStack(getInteger(ctx, "count"), false))))
+                                .executes(ctx -> setStack(ctx.getSource(), getString(ctx, "group"), getInteger(ctx, "index"), getItemStackArgument(ctx, "itemstack").createStack(1, false))))))
+                    .then(literal("icon")
+                        .then(argument("icon", itemStack())
+                            .executes(ctx -> changeIcon(ctx.getSource(), getString(ctx, "group"), getItemStackArgument(ctx, "icon").createStack(1, false)))))
+                    .then(literal("rename")
+                        .then(argument("new", string())
+                            .executes(ctx -> renameGroup(ctx.getSource(), getString(ctx, "group"), getString(ctx, "new")))))))
+            .then(literal("add")
+                .then(argument("group", string())
+                    .then(argument("icon", itemStack())
+                         .executes(ctx -> addGroup(ctx.getSource(), getString(ctx, "group"), getItemStackArgument(ctx, "icon").createStack(1, false))))))
+            .then(literal("remove")
+                .then(argument("group", string())
+                    .suggests((ctx, builder) -> suggestMatching(groups.keySet(), builder))
+                    .executes(ctx -> removeGroup(ctx.getSource(), getString(ctx, "group"))))));
+    }
+
+    private static int addGroup(ServerCommandSource source, String name, ItemStack icon) throws CommandSyntaxException {
+        if (groups.containsKey(name)) {
+            throw ALREADY_EXISTS_EXCEPTION.create(name);
+        }
+
+        try {
+            ItemGroup itemGroup = FabricItemGroupBuilder.create(
+                    new Identifier("clientcommands", name))
+                    .icon(() -> icon)
+                    .build();
+
+            groups.put(name, new Group(itemGroup, icon, new ListTag()));
+            saveFile();
+            sendFeedback("commands.citemgroup.addGroup.success", name);
+        } catch (InvalidIdentifierException e) {
+            throw ILLEGAL_CHARACTER_EXCEPTION.create(name);
+        }
+        return 0;
+    }
+
+    private static int removeGroup(ServerCommandSource source, String name) throws CommandSyntaxException {
+        if (!groups.containsKey(name)) {
+            throw NOT_FOUND_EXCEPTION.create(name);
+        }
+
+        groups.remove(name);
+
+        reloadGroups(12);
```

Comment:
Hardcoding this value 12 seems dangerous.What if other mods add item groups? What if a new Minecraft update adds new item groups? There must be a better way to do this, perhaps by filtering out the clientcommands groups?

Reply:
Yeah this is difficult. If other mods would add an item group, it would be deleted. I will try to implement a check for clientcommands-created item groups. I do not worry as much about a Minecraft update adding a new group, that will just change the starting index (12 now).

Reply:
Does this look reasonable?
```java
@Override
public void shrink(int index) {
    ItemGroup[] tempGroups = GROUPS;
    tempGroups = Arrays.stream(tempGroups).filter(itemGroup -> !itemGroup.getName().startsWith("clientcommands.")).toArray(ItemGroup[]::new);
    GROUPS = new ItemGroup[index + tempGroups.length];
    System.arraycopy(tempGroups, 0, GROUPS, 0, GROUPS.length);
}
```

Reply:
To be honest I'm going quite insane from this mess of arrays and indices.

Reply:
Update to clarify: the most recent commit fixed this issue.